src/device_queue.cc

```diff
@@ -4,282 +4,218 @@
+        handle_( handle_create( streams_[ 0 ] ) ),
+        events_  { nullptr },                  // all events are null
+        num_active_streams_( 1 ),
+        current_stream_index_( 0 ),
+        own_handle_        ( true ),
+        own_default_stream_( true ),
+
+    #elif defined( BLAS_HAVE_ONEMKL )
+        streams_{ sycl::queue( DeviceList::at( device ), s_property_in_order ) },
+
+    #else
+        // No GPU
+        streams_{ nullptr },
+    #endif
+    device_( device )
+{}
+
+#if defined( BLAS_HAVE_CUBLAS ) || defined( BLAS_HAVE_ROCBLAS )
+    // -------------------------------------------------------------------------
+    /// Constructor taking a cuBLAS or rocBLAS handle.
+    /// This gets the stream from the given handle.
+    /// The user retains ownership of the stream and handle,
+    /// which must exist whenever this queue is used.
+    Queue::Queue( int device, handle_t handle )
+      : work_( nullptr ),
+        lwork_( 0 ),
+        streams_ { handle_get_stream( handle ) },  // remaining streams are null
+        handle_( handle ),
+        events_  { nullptr },  // all events  are null
+        num_active_streams_( 1 ),
+        current_stream_index_( 0 ),
+        own_handle_        ( false ),
+        own_default_stream_( false ),
+        device_( device )
+    {}
+
+    // -------------------------------------------------------------------------
+    /// Constructor taking a CUDA or HIP stream.
+    /// This allocates a cublasHandle_t or rocblas_handle,
+    /// and associates the given stream with it.
+    /// The user retains ownership of the stream,
+    /// which must exist whenever this queue is used.
+    Queue::Queue( int device, stream_t& stream )
+      : work_( nullptr ),
+        lwork_( 0 ),
+        streams_ { nullptr },      // remaining streams are null
+        handle_( nullptr ),  // created below due to set_device
+        events_  { nullptr },     // all events  are null
+        num_active_streams_( 1 ),
+        current_stream_index_( 0 ),
+        own_handle_        ( true  ),
+        own_default_stream_( false ),
+        device_( device )
+    {
+        internal_set_device( device_ );
+        streams_[ 0 ] = stream;
+        handle_ = handle_create( stream );
+    }
 
-        // create parallel streams
-        for (size_t i = 0; i < DEV_QUEUE_FORK_SIZE; ++i) {
-            stream_create(&parallel_streams_[ i ]);
+    // -------------------------------------------------------------------------
+    /// Change the CUDA or HIP stream used in the BLAS++ queue.
+    /// Kernels executing on the current stream will continue.
+    /// Throws an error if in fork mode.
+    void Queue::set_stream( stream_t& stream )
+    {
+        if (current_stream_index_ != 0)
+            throw blas::Error( "can't set stream in fork mode", __func__ );
+
+        if (own_default_stream_) {
+            stream_destroy( streams_[ 0 ] );
```

Comment:
Do not we need to check if a handle is associated to the stream and destroy it if so?

Reply:
I don't think so. There's only one handle, and we change that handle to use this stream below.